software/firmware/experimental/custom_font.py

```diff
@@ -0,0 +1,157 @@
+import framebuf
+
+from machine import I2C
+from machine import Pin
+from ssd1306 import SSD1306_I2C
+
+from europi import OLED_WIDTH, I2C_FREQUENCY, OLED_HEIGHT, I2C_CHANNEL, TEST_ENV, CHAR_HEIGHT
+from europi import Display as BasicDisplay
+
+
+# TODO: add a method to select the font to use by default
+
+
+class CustomFontWriter:
+    def __init__(self, device, font):
+        """Initialize the Writer.
+
+        device: the OLED display instance
+        font: a font module
+        """
+        self.device = device
+        self.font = font
+        if font.hmap():
+            self.map = framebuf.MONO_HMSB if font.reverse() else framebuf.MONO_HLSB
+        else:
+            raise ValueError("Font must be horizontally mapped.")
+        self.screenwidth = device.width  # In pixels
+        self.screenheight = device.height
+
+    def print(self, string, x, y, c=1):
+        """Print the string using the x, y coordinates as the upper-left corner of the text.
+        With c=0, the text is display as black text on white background.
+        """
+        for char in string:
+            if char == "\n":  # line breaks are ignored
+                return
+            glyph, char_height, char_width = self.font.get_ch(char)
+            buf = bytearray(glyph)
+            if c != 1:
+                for i, v in enumerate(buf):
+                    buf[i] = 0xFF & ~v
+            fbc = framebuf.FrameBuffer(buf, char_width, char_height, self.map)
+            self.device.blit(fbc, x, y)
+            x += char_width
+
+    def string_len(self, string):
+        """Returns the length of string in pixels."""
+        n = 0
+        for char in string:
+            n += self._char_len(char)
+        return n
+
+    def _char_len(self, char):
+        """Returns the length of char in pixels."""
+        if char == "\n":
+            char_width = 0
+        else:
+            _, _, char_width = self.font.get_ch(char)
+        return char_width
+
+
+class Display(BasicDisplay):
+    """A class for drawing graphics and text to the OLED.
```

Comment:
[required] This docstring is copied from the parent. Instead it should document the features that this class makes available over the parent.